File: go/pkg/parser/func_declaration_visitor.go
package parser

import "github.com/mojo-lang/lang/go/pkg/lang"

type FuncDeclarationVisitor struct {
	*BaseMojoParserVisitor
}

func NewFuncDeclarationVisitor() *FuncDeclarationVisitor {
	decl := &FuncDeclarationVisitor{}
	return decl
}

func (f *FuncDeclarationVisitor) VisitFunctionDeclaration(ctx *FunctionDeclarationContext) interface{} {
	decl := &lang.FuncDecl{}

	nameCtx := ctx.FunctionName()
	if nameCtx != nil {
		decl.Name = nameCtx.Accept(f).(string)
	}

	decl.GenericParameters = GetGenericParameters(ctx.GenericParameterClause())

	signatureCtx := ctx.FunctionSignature()
	if signatureCtx != nil {
		decl.Type = signatureCtx.Accept(f).(*lang.FuncType)
	}

	return decl
}

func (f *FuncDeclarationVisitor) VisitInterfaceMethodDeclaration(ctx *InterfaceMethodDeclarationContext) interface{} {
	decl := &lang.FuncDecl{}
	
	decl.Document = GetDocument(ctx.Document())
	decl.Attributes = GetAttributes(ctx.Attributes())

	nameCtx := ctx.FunctionName()
	if nameCtx != nil {
		decl.Name = nameCtx.Accept(f).(string)
	}

	decl.GenericParameters = GetGenericParameters(ctx.GenericParameterClause())

	signatureCtx := ctx.FunctionSignature()
	if signatureCtx != nil {
		decl.Type = signatureCtx.Accept(f).(*lang.FuncType)
	}

	return decl
}

func (f *FuncDeclarationVisitor) VisitFunctionName(ctx *FunctionNameContext) interface{} {
	return ctx.GetText()
}

func (f *FuncDeclarationVisitor) VisitFunctionSignature(ctx *FunctionSignatureContext) interface{} {
	funcType := &lang.FuncType{}
	
	parameterCtx := ctx.FunctionParameterClause()
	if parameterCtx != nil {
		funcType.Parameters = parameterCtx.Accept(f).([]*lang.ValueDecl)
	}

	resultCtx := ctx.FunctionResult()
	if resultCtx != nil {
		funcType.ReturnType = resultCtx.Accept(f).(*lang.NominalType)
	}

	return funcType
}

func (f *FuncDeclarationVisitor) VisitFunctionParameterClause(ctx *FunctionParameterClauseContext) interface{} {
	paramCtx := ctx.FunctionParameters()
	if paramCtx != nil {
		return paramCtx.Accept(f)
	}

	return []*lang.ValueDecl{}
}

func (f *FuncDeclarationVisitor) VisitFunctionParameters(ctx *FunctionParametersContext) interface{} {
	paramCtxes := ctx.AllFunctionParameter()
	var parameters []*lang.ValueDecl
	for _, paramCtx := range paramCtxes {
		visitor := NewValueDeclarationVisitor()
		value := paramCtx.Accept(visitor).(*lang.ValueDecl)
		if value != nil {
			parameters = append(parameters, value)
		}
	}

	return parameters
}

func (f *FuncDeclarationVisitor) VisitFunctionResult(ctx *FunctionResultContext) interface{} {
	returnType := GetType(ctx.Type_())
	if returnType != nil {
		returnType.Attributes = GetAttributes(ctx.Attributes())
	}

	return returnType
}

func (f *FuncDeclarationVisitor) VisitFunctionBody(ctx *FunctionBodyContext) interface{} {
	return nil
}

File: java/mojo-parser/src/main/java/org/mojolang/mojo/parser/FuncDeclarationVisitor.java
package org.mojolang.mojo.parser;

import com.google.protobuf.GeneratedMessageV3;
import org.mojolang.mojo.lang.FuncDecl;
import org.mojolang.mojo.lang.FuncType;
import org.mojolang.mojo.lang.NominalType;
import org.mojolang.mojo.lang.ValueDecl;

public class FuncDeclarationVisitor extends MojoBaseVisitor<com.google.protobuf.GeneratedMessageV3> {
    private FuncDecl.Builder builder_ = FuncDecl.newBuilder();
    private FuncType.Builder typeBuilder_= FuncType.newBuilder();

    @Override
    public GeneratedMessageV3 visitFunction_declaration(MojoParser.Function_declarationContext ctx) {
        setFunctionName(ctx.function_name());

        if (ctx.generic_parameter_clause() != null) {
            builder_.addAllGenericParameters(Util.visitGenericParameters(ctx.generic_parameter_clause()));
        }

        visitFunction_signature(ctx.function_signature());

        builder_.setType(typeBuilder_);
        return builder_.build();
    }

    @Override
    public GeneratedMessageV3 visitInterface_method_declaration(MojoParser.Interface_method_declarationContext ctx) {
        setFunctionName(ctx.function_name());

        if (ctx.generic_parameter_clause() != null) {
            builder_.addAllGenericParameters(Util.visitGenericParameters(ctx.generic_parameter_clause()));
        }
        visitFunction_signature(ctx.function_signature());

        builder_.setType(typeBuilder_);
        return builder_.build();
    }

    private void setFunctionName(MojoParser.Function_nameContext ctx) {
        builder_.setName(ctx.declaration_identifier().getText());
        //builder_.setStartPosition(Util.getPosition(ctx.declaration_identifier()));
    }

    @Override
    public GeneratedMessageV3 visitFunction_signature(MojoParser.Function_signatureContext ctx) {
        MojoParser.Function_parameter_listContext parametersCtx = ctx.function_parameter_clause().function_parameter_list();
        if (parametersCtx != null) {
            for (MojoParser.Function_parameterContext parameterCtx : parametersCtx.function_parameter()) {
                ValueDeclarationVisitor visitor = new ValueDeclarationVisitor();
                ValueDecl value = (ValueDecl) visitor.visitFunction_parameter(parameterCtx);

                typeBuilder_.addParameters(value);
            }
        }

        visitFunction_result(ctx.function_result());
        return null;
    }

    @Override
    public GeneratedMessageV3 visitFunction_result(MojoParser.Function_resultContext ctx) {
        if (ctx != null) {
            NominalTypeVisitor visitor = new NominalTypeVisitor();
            NominalType nominalType = (NominalType)visitor.visitTypeAndAttributes(ctx.type_(), ctx.attributes());
            typeBuilder_.setReturnType(nominalType);
        }

        return null;
    }

    @Override
    public GeneratedMessageV3 visitFunction_body(MojoParser.Function_bodyContext ctx) {
        return null;
    }
}
